### PR TITLE
Cleanup compiler warnings

### PR DESCRIPTION
**Pull request by amametjanov**
Cleanup compiler warnings

Fixes E3SM-Project/E3SM#3624, E3SM-Project/E3SM#2250
Addresses E3SM-Project/E3SM#2247

[BFB]

### CODE REVIEW

File: components/cam/src/physics/cam/check_energy.F90 [hunk 1]
```diff
@@ -590,7 +590,9 @@ subroutine check_energy_gmean(state, pbuf2d, dtime, nstep)
 !-----------------------------------------------------------------------
 
     ! Copy total energy out of input and output states
+#ifdef CPRCRAY
 !DIR$ CONCURRENT
+#endif
```
Comment (amametjanov) [inline components/cam/src/physics/cam/check_energy.F90:595]: Cray compiler directive that may be needed when porting and optimizing to Cray. `CPRCRAY` macro cleans up Intel compiler messages like
```
/qfs/people/mame117/repos/E3SM-tests/components/cam/src/physics/cam/check_energy.F90(593): remark #5140: Unrecognized directive
!DIR$ CONCURRENT
----------------^
```

File: components/cam/src/physics/cam/cldwat.F90 [hunk 2]
```diff
@@ -1017,13 +1017,16 @@ subroutine findmcnew (lchnk   ,ncol    , &
...
+   !real(r8) heavy, heavym
+   !heavy(a1,a2) = max(0._r8,sign(1._r8,a1-a2))  ! heavyside function
+   !heavym(a1,a2) = max(0.01_r8,sign(1._r8,a1-a2))  ! modified heavyside function
 !
 ! New heavyside functions to perhaps address error growth problems
 !
+#ifdef PERGRO
    heavyp(a1,a2) = a1/(a2+a1+1.e-36_r8)
+#endif
    heavymp(a1,a2) = (a1+0.01_r8*a2)/(a2+a1+1.e-36_r8)
```
Comment (amametjanov) [inline components/cam/src/physics/cam/cldwat.F90:1030]: Avoid
```
/qfs/people/mame117/repos/E3SM-tests/components/cam/src/physics/cam/cldwat.F90(1020): remark #7713: This statement function has not been used.   [HEAVY]
   real(r8) heavy, heavym, a1, a2, heavyp, heavymp
------------^
/qfs/people/mame117/repos/E3SM-tests/components/cam/src/physics/cam/cldwat.F90(1020): remark #7713: This statement function has not been used.   [HEAVYM]
   real(r8) heavy, heavym, a1, a2, heavyp, heavymp
-------------------^
/qfs/people/mame117/repos/E3SM-tests/components/cam/src/physics/cam/cldwat.F90(1020): remark #7713: This statement function has not been used.   [HEAVYP]
   real(r8) heavy, heavym, a1, a2, heavyp, heavymp
-----------------------------------^
```

File: components/homme/src/share/compose_mod.F90 [hunk 3]
```diff
@@ -341,6 +341,9 @@ subroutine compose_query_bufsz(sendsz, recvsz)
     call cedr_query_bufsz(ssz, rsz)
     sendsz = max(sendsz, ssz)
     recvsz = max(recvsz, rsz)
+#else
+    sendsz = 0
+    recvsz = 0
```
Comment (amametjanov) [inline components/homme/src/share/compose_mod.F90:346]:
```
/qfs/people/mame117/repos/E3SM-tests/components/homme/src/share/compose_mod.F90(334): warning #6843: A dummy argument with an explicit INTENT(OUT) declaration is not given an explicit value.   [SENDSZ]
  subroutine compose_query_bufsz(sendsz, recvsz)
---------------------------------^
/qfs/people/mame117/repos/E3SM-tests/components/homme/src/share/compose_mod.F90(334): warning #6843: A dummy argument with an explicit INTENT(OUT) declaration is not given an explicit value.   [RECVSZ]
  subroutine compose_query_bufsz(sendsz, recvsz)
-----------------------------------------^
```

File: components/cam/src/physics/cam/nudging.F90 [hunk 4]
```diff
@@ -1177,8 +1177,7 @@ subroutine nudging_init
               call alloc_err(istat,'nudging_init','INTP_PS',2*pcols*((endchunk-begchunk)+1))
            end if
       case default
-           call endrun('nudging_init error: nudge method should &
-                        be either Step, Linear or IMT...')
+           call endrun('nudging_init error: nudge method should be either Step, Linear or IMT...')
```
Comment (amametjanov) [inline components/cam/src/physics/cam/nudging.F90:1180]:
```
"/autofs/nccs-svm1_home1/azamat/repos/E3SM-tests/components/cam/src/physics/cam/nudging.F90", line 1181.1: 1515-021 (E) Syntax error: Token " & " is expected.
```

File: components/cam/src/physics/clubb/advance_clubb_core_module.F90 [hunk 5]
```diff
@@ -1612,7 +1612,7 @@ gamma_Skw_fnc, & ! Gamma as a function of skewness          [???]
             rtm_pert_neg_rt = pdf_params_frz%rt_2 & 
                        - Lscale_pert_coef * sqrt( max( pdf_params_frz%varnce_rt_2, rt_tol**2 ) )
             !Lscale_weight = pdf_params%mixt_frac
-          else where
+          elsewhere
```
Comment (amametjanov) [inline components/cam/src/physics/clubb/advance_clubb_core_module.F90:1615]:
```
"/autofs/nccs-svm1_home1/azamat/repos/E3SM-tests/components/cam/src/physics/clubb/advance_clubb_core_module.F90", line 1615.16: 1511-077 (E) ELSEWHERE statement is in error.  ELSEWHERE statement is assumed.
"/autofs/nccs-svm1_home1/azamat/repos/E3SM-tests/components/cam/src/physics/clubb/advance_clubb_core_module.F90", line 1637.16: 1511-077 (E) ELSEWHERE statement is in error.  ELSEWHERE statement is assumed.
```

File: components/clm/src/main/clmfates_paraminterfaceMod.F90 [hunk 6]
```diff
@@ -188,7 +188,7 @@ subroutine ParametersFromNetCDF(filename, is_host_file, fates_params)
    logical, intent(in) :: is_host_file
    class(fates_parameters_type), intent(inout) :: fates_params
 
-   character(len=32)  :: subname = 'clmfates_interface::ReadParameters'
+   character(len=40)  :: subname = 'clmfates_interface::ReadParameters'
```
Comment (amametjanov) [inline components/clm/src/main/clmfates_paraminterfaceMod.F90:191]:
```
"/autofs/nccs-svm1_home1/azamat/repos/E3SM-tests/components/clm/src/main/clmfates_paraminterfaceMod.F90", line 191.26: 1514-038 (W) The literal string has length 34 which is longer than 32, the length of the variable being initialized.  The string will be truncated.
```

File: components/clm/src/main/initGridCellsMod.F90 [hunk 7]
```diff
@@ -1178,6 +1178,7 @@ subroutine CheckGhostSubgridHierarchy()
     enddo
 
   end subroutine CheckGhostSubgridHierarchy
-#endif ! #ifdef USE_PETSC_LIB
+#endif
+!^ifdef USE_PETSC_LIB
```
Comment (amametjanov) [inline components/clm/src/main/initGridCellsMod.F90:1182]:
```
"/autofs/nccs-svm1_home1/azamat/repos/E3SM-tests/components/clm/src/main/initGridCellsMod.F90", line 1181.8: 1506-234 (W) Expecting a new-line character on #endif directive.
```

**Comment by amametjanov** [comment]
A related PR in FATES submodule: NGEET/fates#666

File: components/cam/src/chemistry/mozart/wei96.F90 [hunk 8]
```diff
@@ -926,7 +927,9 @@ INTEGER FUNCTION JULDAY(MM,ID,IYYY)
 !
 !-----------------------------------------------------------------------
 !
-      IF (IYYY.EQ.0) PAUSE 'There is no Year Zero.'
+      IF (IYYY.EQ.0) THEN
+        call endrun('There is no Year Zero.')
+      ENDIF
```
Comment (amametjanov) [inline components/cam/src/chemistry/mozart/wei96.F90:932]:
```
/lcrc/group/acme/azamat/repos/E3SM-tests/components/cam/src/chemistry/mozart/wei96.F90:929:51:

       IF (IYYY.EQ.0) PAUSE 'There is no Year Zero.'
                                                   1
Warning: Deleted feature: PAUSE statement at (1)
```

File: components/cice/src/mpi/ice_boundary.F90 [hunk 9]
```diff
@@ -2925,7 +2925,7 @@ subroutine ice_HaloUpdate3DR8(array, halo,                    &
       ioffset, joffset,          &! address shifts for tripole
       isign                       ! sign factor for tripole grids
 
-   integer (int_kind), dimension(:), allocatable, save :: &
+   integer (int_kind), dimension(:), allocatable :: &
       sndRequest,      &! MPI request ids
       rcvRequest        ! MPI request ids
```
Comment (amametjanov) [inline components/cice/src/mpi/ice_boundary.F90:2930]:
```
/lcrc/group/acme/azamat/repos/E3SM-tests/components/cice/src/mpi/ice_boundary.F90:2958:22:

       bufSend, bufRecv            ! 3d send,recv buffers
                      1
Warning: Legacy Extension: Duplicate SAVE attribute specified at (1)
/lcrc/group/acme/azamat/repos/E3SM-tests/components/cice/src/mpi/ice_boundary.F90:2958:13:

       bufSend, bufRecv            ! 3d send,recv buffers
             1
Warning: Legacy Extension: Duplicate SAVE attribute specified at (1)
/lcrc/group/acme/azamat/repos/E3SM-tests/components/cice/src/mpi/ice_boundary.F90:2961:16:

       bufTripole                  ! 3d tripole buffer
                1
Warning: Legacy Extension: Duplicate SAVE attribute specified at (1)
/lcrc/group/acme/azamat/repos/E3SM-tests/components/cice/src/mpi/ice_boundary.F90:2930:16:

       rcvRequest        ! MPI request ids
                1
Warning: Legacy Extension: Duplicate SAVE attribute specified at (1)
/lcrc/group/acme/azamat/repos/E3SM-tests/components/cice/src/mpi/ice_boundary.F90:2929:16:

       sndRequest,      &! MPI request ids
                1
Warning: Legacy Extension: Duplicate SAVE attribute specified at (1)
```

File: components/cice/src/source/ice_forcing.F90 [hunk 10]
```diff
@@ -3248,12 +3248,12 @@ subroutine ocn_data_ncar_init
...
 ! 1  sst------temperature---------------------------(C)   \\
 ! 2  sss------salinity------------------------------(ppt) \\
-! 3  hbl------depth---------------------------------(m)   \\ 
+! 3  hbl------depth---------------------------------(m)   \\
 ! 4  u--------surface u current---------------------(m/s) \\
 ! 5  v--------surface v current---------------------(m/s) \\
 ! 6  dhdx-----surface tilt x direction--------------(m/m) \\
 ! 7  dhdy-----surface tilt y direction--------------(m/m) \\
-! 8  qdp------ocean sub-mixed layer heat flux-------(W/m2)\\ 
+! 8  qdp------ocean sub-mixed layer heat flux-------(W/m2)\\
```
Comment (amametjanov) [inline components/cice/src/source/ice_forcing.F90:3256]:
```
/lcrc/group/acme/azamat/repos/E3SM-tests/components/cice/src/source/ice_forcing.F90:3251:60:

 ! 3  hbl------depth---------------------------------(m)   \\

Warning: backslash and newline separated by space
/lcrc/group/acme/azamat/repos/E3SM-tests/components/cice/src/source/ice_forcing.F90:3256:60:

 ! 8  qdp------ocean sub-mixed layer heat flux-------(W/m2)\\

Warning: backslash and newline separated by space
```

File: components/clm/src/biogeophys/WaterBudgetMod.F90 [hunk 11]
```diff
@@ -105,8 +105,6 @@ module WaterBudgetMod
   real(r8) :: budg_stateL(s_size, p_size)
   real(r8), public :: budg_stateG(s_size, p_size)
 
-  logical,save :: first_time = .true.
-
```
Comment (amametjanov) [inline components/clm/src/biogeophys/WaterBudgetMod.F90:109]:
```
/lcrc/group/acme/azamat/repos/E3SM-tests/components/clm/src/biogeophys/WaterBudgetMod.F90:108:28:

   logical,save :: first_time = .true.
                            1
Warning: Legacy Extension: Duplicate SAVE attribute specified at (1)
```
This var isn't used anywhere.

File: components/homme/src/share/compose_mod.F90 (hunk 3 above)
Comment (singhbalwinder) [reply components/homme/src/share/compose_mod.F90:346]: I haven't looked at the code to know how these variables are being used. It is probably okay but @ambrad or @oksanaguba might know if it is okay to assign these to zero.

**Comment by singhbalwinder** [comment]
Do we know if the CRAY specific directives are still useful? If not, we can simply remove them.

File: components/homme/src/share/compose_mod.F90 (hunk 3 above)
Comment (ambrad) [reply components/homme/src/share/compose_mod.F90:346]: Yes, these lines are fine. The subroutine isn't even called in this branch, but it's good to take care of compiler warnings. Thanks! (And thanks for tagging me @singhbalwinder.)

**Comment by amametjanov** [comment]
Someone identified these loops for optimization and these will be good starting points when optimizing for newer Cray compiler versions.

**Comment by AaronDonahue** [comment]
It looks like this PR is ready to go, I'm going to start the integration process.

**Pull request merged**